_posts/2024-08-07-accessing-kafka-with-gateway-api.md

```diff
@@ -0,0 +1,627 @@
+---
+layout: post
+title: "Accessing Kafka with the Gateway API"
+date: 2024-08-07
+author: colt_mcnealy
+---
+
+At [LittleHorse](https://littlehorse.dev), we use the [Gateway API](https://gateway-api.sigs.k8s.io) to allow external traffic into our kubernetes clusters. We will soon need to allow external clients to access our [Kafka](https://kafka.apache.org) clusters (managed by Strimzi, of course!) from outside of our Kubernetes clusters. We have so far been pleased with the performance and simplicity of [Envoy Gateway](https://gateway.envoyproxy.io/) as a Gateway Controller, which motivated this investigation into using Envoy Gateway to access our Kafka clusters.
+
+## Background
+
+The [Gateway API](https://gateway-api.sigs.k8s.io/) in Kubernetes aims to replace the `Ingress` resource as the de facto standard for allowing external traffic to reach workloads running on Kubernetes. It addresses many shortcomings of the `Ingress` resource, including poor support for non-HTTP 1.0 traffic.
+
+![Strimzi and Envoy Gateway](/assets/images/posts/2024-08-07-strimzi-gateway-api.png)
+
+### Accessing Kafka
+
+Previously, Jakub Scholz blogged about how Strimzi allows you to access Kafka from outside the Kubernetes cluster using [`NodePort` services](https://strimzi.io/2019/04/23/accessing-kafka-part-2.html), [OpenShift `Route`s](https://strimzi.io/2019/04/30/accessing-kafka-part-3.html), [`LoadBalancer` Services](https://strimzi.io/2019/05/13/accessing-kafka-part-4.html), and [`Ingress` resources](https://strimzi.io/blog/2019/05/23/accessing-kafka-part-5/).
+
+As Jakub [noted](https://strimzi.io/blog/2019/04/17/accessing-kafka-part-1/), accessing Kafka is difficult for two reasons:
+
+1. Kafka Clients need to be able to access specific brokers individually, so simply scattering the requests across the Kafka Cluster using a load balancer would yield incorrect results.
+2. The Kafka protocol is not based on HTTP, which means that you need a few clever hacks to get it to work with plain `Ingress`.
+
+### The Gateway API
+
+The Gateway API is a much more flexible and extensible solution for north-south traffic than `Ingress`. The entirety of the Gateway API is beyond the scope of this post, but there are two resources in particular that will be of interest to us:
+
+1. `TCPRoute`s, which proxy unencrypted TCP traffic.
+2. `TLSRoute`s, which control encrypted TCP traffic.
+
+The `HTTPRoute` and `GRPCRoute` resources will not work with Kafka because Kafka does not speak an HTTP-based protocol.
```

Comment:
Maybe this will make it a bit more clear for the readers?

```suggestion
The `HTTPRoute` and `GRPCRoute` resources will not work with Kafka because Kafka does not speak an HTTP or GRPC-based protocol.
```